### ISSUE TITLE
Docker daemon source transport layer doesn't check for existing blob in case of oci destination

### ISSUE THREAD

**Issue by praveenkumar** [containers/image#475]
Is there any specific reason behind this?

```
$ skopeo copy docker-daemon:fedora:latest oci:fedora
Getting image source signatures
Copying blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
 250.51 MB / 250.51 MB [===================================================] 16s
Copying config sha256:5415dde2eeab8b18b37d43db94983e3c8f5e3f79c9c5f64ded360ae0c38204d6
 1.31 KB / 1.31 KB [========================================================] 0s

$ skopeo copy docker-daemon:fedora:latest oci:fedora <= run it again and this doesn't check for existing blob but just start copy.
Getting image source signatures
Copying blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
 250.51 MB / 250.51 MB [===================================================] 16s
Copying config sha256:5415dde2eeab8b18b37d43db94983e3c8f5e3f79c9c5f64ded360ae0c38204d6
 1.31 KB / 1.31 KB [========================================================] 0s
Writing manifest to image destination

$ sudo skopeo copy docker-daemon:fedora:latest containers-storage:fedora <= works as expected.
Getting image source signatures
Skipping fetch of repeat blob sha256:891e1e4ef82ad02a4ea1f030831f942d722c7694c4db64ca3239c8163b811c58
Copying config sha256:cc510acfcd701a409014118d5f417f0022520802a26c650866b8a9594d75f3a7
 2.29 KB / 2.29 KB [========================================================] 0s
Writing manifest to image destination
```

**Comment by mtrmac** [comment]
“Yes” :)

Writing to OCI compresses the layers, which makes it impossible to check whether the original uncompressed layer is present when doing a second copy.

**Comment by praveenkumar** [comment]
@mtrmac Is there any option to put un-compresses layer when using with oci destination or it is just a limitation for us right now?

**Comment by mtrmac** [comment]
There isn't one right now.  What do you ultimately _need to do_?

**Comment by praveenkumar** [comment]
> There isn't one right now. What do you ultimately need to do?

@mtrmac As part of minishift [0] we use oci as the destination source to cache the docker images using `minishift image export` (https://github.com/minishift/minishift/blob/master/pkg/minishift/docker/image/oci_image_handler.go#L311-L340) and this works with all the different platform (win, mac and linux) but as you know performance affected if there is some layers already exist but then also it again download/copy all :(

[0] https://github.com/minishift/minishift

**Comment by vrothberg** [comment]
It works when using docker instead of docker-daemon as the source. The docker-daemon has some limitations compared to other transports (e.g., using it as a destination will always copy).

**Comment by mtrmac** [comment]
> It works when using docker instead of docker-daemon as the source.

In _this_ aspect, It works pretty much the same in both cases - if a `docker:` source contained uncompressed blobs (which, to be fair, pretty much never happens), `skopeo` would not detect the duplicate either.

> The docker-daemon has some limitations compared to other transports (e.g., using it as a destination will always copy).

Yes, that performance impact is always there. IIRC a `docker-daemon:` source creates _two_ on-disk copies of the data before `skopeo` even gets access to it (and makes a third one); and only after all this disk activity can some duplicate detection happen.

**Comment by mtrmac** [comment]
> As part of minishift [0] we use oci as the destination source to cache the docker images using minishift image export

Thanks.

Overall, I do realize that use cases are what they are, and I can’t dictate that the world should change to make life easier for this project; nevertheless let me at least _suggest_ to think whether it would be possible to reorganize the code to minimize usage of the Docker daemon’s container storage as an image store.

As you see here, that comes with fairly noticeable performance impact of several kinds, some not reasonably avoidable (the multiple on-disk copies), some maybe avoidable (the recompression discussed in this issue).

Much worse, though, is that the storage is inherently lossy; the original manifest is dropped, and the layer contents are decompressed and re-compressed, making them different in the general case.  That makes it more difficult to determine whether an image has been modified (or at the very least, it forces the trusted image authentication code to run risky decompression before authenticating the layer contents).

E.g., maybe https://github.com/minishift/minishift/blob/f84f63281357a9437849e96e6547a3f15336a2fa/pkg/minishift/docker/image/oci_image_handler.go#L317 could be reworked, so that instead of `docker pull` + `docker save`, it does a direct `containers/image/copy.Image` from a `docker://` source to the `oci:` cache, and _then_, if necessary, an “import” into the daemon.

(That would, currently, still not preserve the manifest due to a v2s2→OCI conversion, but at least the compressed layer data would be copied as-is; and as soon as the registries start publishing OCI-formatted images, the original manifests will be preserved.)

---

As for what can be done with the use case as is: It would be easy enough to allow callers to configure the OCI backend so that it does not compress layers.  That would “solve” the issue (i.e. the code would do “only” 3 uncompressed on-disk copies and have a working duplicate detection, instead of the current state of 3 uncompressed on-disk copies + compression load on CPU + 4th on-disk compressed copy) — at the cost of notably larger use of disk space. It’s up to you whether this would be a worthwhile change; I’m skeptical.

---

Ultimately, we want `skopeo` to cache the known compressed/uncompressed layer equivalencies, so that it can do the duplicate detection in such cases as well (https://github.com/projectatomic/skopeo/issues/395 ); while that is primarily aimed at `docker://` destinations, it might be applicable to `oci:` as well.  But I am not sure when anyone will have time to work on this.

**Comment by praveenkumar** [comment]
@mtrmac Thank you for the explanation. 

> Could be reworked, so that instead of docker pull + docker save, it does a direct containers/image/copy.Image from a docker:// source to the oci: cache, and then, if necessary, an “import” into the daemon.

https://github.com/praveenkumar/minishift/blob/45ef7845c5ad2eb391eeac90e28fc73b5f42632b/pkg/minishift/docker/image/oci_image_handler.go#L325-L349 is what I also did as part of the PR instead of directly consuming the docker-daemon let's consume `docker` as source but in user is offline then `docker` source is unreachable so fall back to `docker-daemon` as source. 

> That would, currently, still not preserve the manifest due to a v2s2→OCI conversion, but at least the compressed layer data would be copied as-is; and as soon as the registries start publishing OCI-formatted images, the original manifests will be preserved.

Right but that is still not the case.

> As for what can be done with the use case as is: It would be easy enough to allow callers to configure the OCI backend so that it does not compress layers. That would “solve” the issue (i.e. the code would do “only” 3 uncompressed on-disk copies and have a working duplicate detection, instead of the current state of 3 uncompressed on-disk copies + compression load on CPU + 4th on-disk compressed copy) — at the cost of notably larger use of disk space. It’s up to you whether this would be a worthwhile change; I’m skeptical.

I think that something all the consumers of this library decide (skepeo, podman, minishift ...etc.) If that solve problem without any regression.

**Comment by jorgemoralespou** [comment]
@mtrmac as I read this, it seems as OCI storage might not be the most
appropriate one for/when working on this use case, and potentially a
containers-storage backend would be much more beneficial in terms of speed
and ease, but that seem not to be available for Windows and macos hosts.
Would it be too complex to work on those backend? Bare with me I'm not a
technical guy so maybe I'm saying something stupid.

**Comment by mtrmac** [comment]
> > As for what can be done with the use case as is: It would be easy enough to allow callers to configure the OCI backend so that it does not compress layers. That would “solve” the issue (i.e. the code would do “only” 3 uncompressed on-disk copies and have a working duplicate detection, instead of the current state of 3 uncompressed on-disk copies + compression load on CPU + 4th on-disk compressed copy) — at the cost of notably larger use of disk space. It’s up to you whether this would be a worthwhile change; I’m skeptical.
>
> I think that something all the consumers of this library decide (skepeo, podman, minishift ...etc.) If that solve problem without any regression.

(To be explicit, this would be an option passed to the library; all the consumers do not have to agree on a single value.)

Well, not a “regression”, but there is certainly a _time/space tradeoff_.  A quick test on my personal system:

With the current behavior (compressing layers, broken duplicate check), both the first and any subsequent copy takes 14.6 seconds, and the resulting directory (for a `fedora` image) is 83 MB large.

Without compression (`DesiredLayerCompression` returning `PreserveOriginal`), the first copy takes 4.3 seconds, the subsequent ones 2.7 seconds — _but_ the directory size is 251 MB.

(It turns out that the copy is actually overwhelmingly CPU-bound while compressing the layers, at least on my system. It might be different with a rotational hard drive.)

Are you at all interested in a 12-second speedup at the cost of using 200% more storage?

(With the speculated compressed/uncompressed equivalency cache, we could probably achieve 14.6 seconds for a first copy, 2.7 seconds for subsequent ones, and 83 MB size. But that’s not too likely to happen very soon, unless you’d care to contribute that.)

**Comment by mtrmac** [comment]
> @mtrmac as I read this, it seems as OCI storage might not be the most
appropriate one for/when working on this use case, and potentially a
containers-storage backend would be much more beneficial in terms of speed
and ease

No; the difference in duplicate detection is _entirely_ driven by whether the backend asks for layer compression or not.  (`containers-storage:` is inherently uncompressed; `oci:` can in principle support both, but right now always asks for compression).

As purely a layer storage mechanism, `containers-storage:` has _much_ more complexity and overhead than `oci:` (it actually extracts the individual files of the layer tarballs, in some fashion; that involves system calls to the kernel writing directory entries, and all the disk seeks (and journaling) necessary for that).

> but that seem not to be available for Windows and macos hosts.

It may be possible to configure `containers-storage:` to use a non-default driver which does work on those systems, I have no idea. But, just as a replacement for uncompressed `oci:`, that’s not at all attractive.

---

The `containers-storage:` transport _could_, eventually, be more interesting if Minishift stopped using the Docker daemon entirely, and converted to CRI-O and/or podman.  Then the `containers-storage:` transport could be used to directly access the storage underlying CRI-O/podman , a bit similar to the `docker-daemon:` transport, but _much_ more efficient (actually working a single layer at a time instead of making three on-disk copies of the whole image—OTOH the `containers-storage:` store works a file at a time, so it may be slower on rotational media). That might even make the separate `oci:` cache unnecessary; everything could share a single `containers-storage:` store, accessible with reasonable latency.  (For transfers between computers, there would still be the tradeoff between transferring large uncompressed data or spending CPU the to compress it and make it smaller.)

**Comment by praveenkumar** [comment]
> (To be explicit, this would be an option passed to the library; all the consumers do not have to agree on a single value.)

@mtrmac right, that would help all the consumers.

> Are you at all interested in a 12-second speedup at the cost of using 200% more storage?

@jorgemoralespou what is your opinion around this? ATM I will try that and see how our overall cache size effect due to this.

> (With the speculated compressed/uncompressed equivalency cache, we could probably achieve 14.6 seconds for a first copy, 2.7 seconds for subsequent ones, and 83 MB size. But that’s not too likely to happen very soon, unless you’d care to contribute that.)

At that moment, I can't promise to take this up but I am interested to work on that flag which allows consumers to get layers without compression.

**Comment by mtrmac** [comment]
> ATM I will try that and see how our overall cache size effect due to this.

FWIW, to prototype this, hard-code `ociImageDestination` to return `types.PreserveOriginal`.

(A full fix would make it configurable via `types.SystemContext`, the way `dirImageDestination` does it.)

**Comment by praveenkumar** [comment]
> FWIW, to prototype this, hard-code ociImageDestination to return types.PreserveOriginal.
(A full fix would make it configurable via types.SystemContext, the way dirImageDestination does it.)

@mtrmac thanks, I will take a look into this tomorrow and try to send PR soon.

**Comment by jorgemoralespou** [comment]
>> Are you at all interested in a 12-second speedup at the cost of using 200% more storage?

> @jorgemoralespou what is your opinion around this? ATM I will try that and see how our overall cache size effect due to this.

@praveenkumar I think that it's something that could be interesting as long as we provide a nice "pruning" mechanism for the cache as it gets bigger.

Most of the images in origin should share same base layers, so ideally this would not add that much overhead on disk space, as long as layers are properly re-used.

**Comment by praveenkumar** [comment]
> Most of the images in origin should share same base layers, so ideally this would not add that much overhead on disk space, as long as layers are properly re-used.

@jorgemoralespou so I made changes as per @mtrmac suggested and also send PR. now below results from the minishift side. you can notice there is a good gain in time but also we are now using more storage.

With that PR

```
$ time ./minishift image export --all
Exporting 'docker.io/openshift/origin-deployer:latest'............. OK
Exporting 'docker.io/openshift/origin-docker-registry:latest'........... OK
Exporting 'docker.io/openshift/origin-haproxy-router:latest'............ OK
Exporting 'docker.io/openshift/origin-pod:latest'......... OK
Exporting 'docker.io/openshift/origin:latest'..................... OK

real	2m31.220s
user	0m15.846s
sys	0m22.112s

$ du -sh ~/.minishift/cache/images/
961M	/home/prkumar/.minishift/cache/images/
```

Without PR

```
$ time ./minishift image export --all
Exporting 'docker.io/openshift/origin-deployer:latest'.................... OK
Exporting 'docker.io/openshift/origin-docker-registry:latest'................. OK
Exporting 'docker.io/openshift/origin-haproxy-router:latest'..................... OK
Exporting 'docker.io/openshift/origin-pod:latest'................ OK
Exporting 'docker.io/openshift/origin:latest'.................................. OK

real	4m54.897s
user	2m16.110s
sys	0m30.679s

$ du -sh ~/.minishift/cache/images/
249M	/home/prkumar/.minishift/cache/images/
```

**Comment by rhatdan** [comment]
BTW I will take speed over size everytime.  We can add features like ostree, reflink and other stuff to help with the size issues.

**Comment by mtrmac** [comment]
OSTree / reflink can’t deduplicate an extracted layer against the tarball; the Minishift usage of this OCI cache amounts to storing two copies of the uncompressed images on the same system. Any deduplication that happens within the OCI repo works, more or less, just the same for compressed and uncompressed data, and does not change the relative size increase.

(It may well make sense to use the uncompressed storage; just don’t expect OSTree to make things better.)

**Comment by rhatdan** [comment]
@praveenkumar @mtrmac @vrothberg I believe we have support for this in buidlah and podman now, can I close this issue?

**Comment by mtrmac** [comment]
There is a `types.SystemContext.OCIAcceptUncompressedLayers` now, and Minishift uses it.

The contemplated blob info cache has also been implemented (in #536 ), but it does not help with this case, because we OCI’s `TryReusingBlob` does not use it. That _could_ be done, but I’m not at all sure it’s a good idea: local-file-path sources/destinations are often enough short-term temporary directories; and recording such locations in the blob info cache would bloat it with little benefit.

So, closing until a stronger need for OCI local-file support for the blob info cache emerges. @praveenkumar please reopen this if it would noticeably help your use case.
